lib/rules/color-hex-alpha/index.js

```diff
@@ -0,0 +1,68 @@
+// @ts-nocheck
+'use strict';
+
+const declarationValueIndex = require('../../utils/declarationValueIndex');
+const report = require('../../utils/report');
+const ruleMessages = require('../../utils/ruleMessages');
+const validateOptions = require('../../utils/validateOptions');
+const valueParser = require('postcss-value-parser');
+
+const ruleName = 'color-hex-alpha';
+
+const messages = ruleMessages(ruleName, {
+	expected: (hex) => `Expected alpha channel in "${hex}"`,
+	unexpected: (hex) => `Unexpected alpha channel in "${hex}"`,
+});
+
+const HEX = /^#([\da-f]{3,4}|[\da-f]{6}|[\da-f]{8})$/i;
+
+function rule(primary) {
+	return (root, result) => {
+		const validOptions = validateOptions(result, ruleName, {
+			actual: primary,
+			possible: ['always', 'never'],
+		});
+
+		if (!validOptions) return;
+
+		root.walkDecls((decl) => {
+			const parsedValue = valueParser(decl.value);
+
+			parsedValue.walk((node) => {
+				if (isUrlFunction(node)) return false;
+
+				if (!isHexColor(node)) return;
+
+				const { value } = node;
+
+				if (primary === 'always' && hasAlphaChannel(value)) return;
+
+				if (primary === 'never' && !hasAlphaChannel(value)) return;
+
+				report({
+					message: primary === 'never' ? messages.unexpected(value) : messages.expected(value),
+					node: decl,
+					index: declarationValueIndex(decl) + node.sourceIndex,
+					result,
+					ruleName,
+				});
+			});
+		});
+	};
+}
+
+function isUrlFunction({ type, value }) {
+	return type === 'function' && value === 'url';
+}
+
+function isHexColor({ type, value }) {
+	return type === 'word' && HEX.test(value);
+}
```

Comment:
[question] There are similar functions named `isHexColor` in other rules. Does it make sense to extract a utility function, like `lib/utils/isHexColor.js`? (maybe out of the scope of this PR)

https://github.com/stylelint/stylelint/blob/b01ed25dfa3e8231a976eef76bd5e81fb535b1e9/lib/rules/color-hex-case/index.js#L76-L78

https://github.com/stylelint/stylelint/blob/b01ed25dfa3e8231a976eef76bd5e81fb535b1e9/lib/rules/color-hex-length/index.js#L114-L116

Reply:
Good idea, but one for a refactor pull request.

We'll need to take into account that `color-hex-case` should check anything hex-like, whereas the other rules should only work on valid hex colours.

Reply:
Thanks, I see that regexes are slightly different for each rule with reason.
It seems reasonable to keep them as-is.